gapii/cc/to_proto.h

```diff
@@ -23,141 +23,131 @@
-        in.toProto(out);
+struct ProtoConverter<typename T::ProtoType, T> {
+    static inline void convert(typename T::ProtoType* out, const T& in, ToProtoContext& ctx) {
+        in.toProto(out, ctx);
     }
 };
 
-// core::StaticArray<In, N> -> ::google::protobuf::RepeatedPtrField<Out>*
+// Converts StaticArray to proto
 template <typename Out, typename In, int N>
-struct ProtoConverter<::google::protobuf::RepeatedPtrField<Out>*, core::StaticArray<In, N>> {
-    static inline void convert(::google::protobuf::RepeatedPtrField<Out>* out, const core::StaticArray<In, N>& in) {
+struct ProtoConverter<Out, core::StaticArray<In, N>> {
+    static inline void convert(Out* out, const core::StaticArray<In, N>& in, ToProtoContext& ctx) {
         out->Reserve(N);
         for (int i = 0; i < N; i++) {
-            ProtoConverter<Out*, In>::convert(out->Add(), in[i]);
+            toProto(out->Add(), in[i], ctx);
         }
     }
 };
 
-// core::StaticArray<In, N> -> ::google::protobuf::RepeatedField<Out>*
-template <typename Out, typename In, int N>
-struct ProtoConverter<::google::protobuf::RepeatedField<Out>*, core::StaticArray<In, N>> {
-    static inline void convert(::google::protobuf::RepeatedField<Out>* out, const core::StaticArray<In, N>& in) {
-        out->Reserve(N);
-        for (int i = 0; i < N; i++) {
-            ProtoConverter<Out, In>::convert(out->Add(), in[i]);
+// Converts SharedMap to proto
+template <typename Out, typename K, typename V>
+struct ProtoConverter<Out, SharedMap<K, V>> {
+    static inline void convert(Out* out, const SharedMap<K, V>& in, ToProtoContext& ctx) {
+        auto ref = ctx.GetReferenceID(in.get());
+        out->set_referenceid(ref.first);
+        if (ref.second) {
+            std::map<K, V> sorted(in.begin(), in.end());
+            auto keys = out->mutable_keys();
+            auto values = out->mutable_values();
+            keys->Reserve(sorted.size());
+            values->Reserve(sorted.size());
+            for (auto it : sorted) {
+                toProto(keys->Add(), it.first, ctx);
+                toProto(values->Add(), it.second, ctx);
+            }
+            trimKeys<std::is_integral<K>::value>(keys);
         }
     }
-};
-
-template <typename EntryOut, typename KeyOut, typename KeyIn>
-struct ProtoMapKeyConverter {
-    static inline void convert(EntryOut* out, const KeyIn& in) {
-        KeyOut key;
-        ProtoConverter<KeyOut, KeyIn>::convert(&key, in);
-        out->set_key(key);
-    }
-};
 
-template <typename EntryOut, typename KeyIn>
-struct ProtoMapKeyConverter<EntryOut, typename KeyIn::ProtoType, KeyIn> {
-    static inline void convert(EntryOut* out, const KeyIn& in) {
-        in.toProto(out->mutable_key());
+    // Remove tailing consecutive keys to save space (only applicable for integral keys).
+    template<bool IsInt, typename T>
+    static typename std::enable_if<IsInt>::type trimKeys(T keys) {
```

Comment:
If you are going to go through the fun of using std::enable_if, you can simplify this as

```
template<typename T>
static typename std::enable_if<std::is_integral<T>>::type trimKeys(T keys) {
....
```

Or, if every compiler we target has C++14 support, you can use enable_if_t<>